File: prims.cpp
#include <iostream>
#include <vector>
#include <queue>

typedef std::pair<int, int> pii; // Pair of (weight, vertex)

void prim(const std::vector<std::vector<pii>>& graph) {
    int start_vertex = 0; // Select any starting vertex
    int n = graph.size();

    std::vector<bool> visited(n, false);
    std::priority_queue<pii, std::vector<pii>, std::greater<pii>> pq;
    pq.push(std::make_pair(0, start_vertex));

    while (!pq.empty()) {
        int weight = pq.top().first;
        int vertex = pq.top().second;
        pq.pop();

        if (visited[vertex])
            continue;

        visited[vertex] = true;
        std::cout << "Edge: " << vertex << " - " << vertex << ", Cost: " << weight << std::endl;

        for (const auto& neighbor : graph[vertex]) {
            int next_vertex = neighbor.second;
            int next_weight = neighbor.first;
            if (!visited[next_vertex])
                pq.push(std::make_pair(next_weight, next_vertex));
        }
    }
}

// Example usage
int main() {
    int n; // Number of vertices
    std::cout << "Enter the number of vertices: ";
    std::cin >> n;

    std::vector<std::vector<pii>> graph(n);
    int edges;
    std::cout << "Enter the number of edges: ";
    std::cin >> edges;

    for (int i = 0; i < edges; i++) {
        int u, v, weight;
        std::cout << "Enter edge " << i+1 << " (source, destination, weight): ";
        std::cin >> u >> v >> weight;
        graph[u].push_back(std::make_pair(weight, v));
        graph[v].push_back(std::make_pair(weight, u));
    }

    prim(graph);

    return 0;
}

File: prims.py
import heapq

def prim(graph):
    start_vertex = 0  # Select any starting vertex
    n = len(graph)

    visited = [False] * n
    pq = [(0, start_vertex)]
    heapq.heapify(pq)

    while pq:
        weight, vertex = heapq.heappop(pq)

        if visited[vertex]:
            continue

        visited[vertex] = True

        for neighbor, next_weight in graph[vertex]:
            if not visited[neighbor]:
                heapq.heappush(pq, (next_weight, neighbor))
                print(f"Edge: {vertex} - {neighbor}, Cost: {next_weight}")

# Example usage
n = int(input("Enter the number of vertices: "))
graph = [[] for _ in range(n)]

edges = int(input("Enter the number of edges: "))
for _ in range(edges):
    input_str = input("Enter edge (source, destination, weight): ")
    u, v, weight = map(int, input_str.split())
    graph[u].append((v, weight))
    graph[v].append((u, weight))

prim(graph)


#vertices - 4
#edges - 5
# 0 1 2
# 0 2 4
# 1 3 3
# 2 3 1
# 2 1 5